### PR TITLE
feat: open apps in tabs

### PR DESCRIPTION
**Pull request by Xazin**
Closes: #2942 Relates: #2312

### Feature Preview

https://github.com/AppFlowy-IO/AppFlowy/assets/42929161/c070cf3f-5187-4ea9-84cf-5eddfe689146

---

#### PR Checklist

- [x] My code adheres to the [AppFlowy Style Guide](https://appflowy.gitbook.io/docs/essential-documentation/contribute-to-appflowy/software-contributions/submitting-code/style-guides)
- [x] I've listed at least one issue that this PR fixes in the description above.
- [x] I've added a test(s) to validate changes in this PR, or this PR only contains semantic changes.
- [x] All existing tests are passing.

### CODE REVIEW

File: frontend/appflowy_flutter/lib/workspace/application/tabs/tabs_state.dart [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+    _pageManagers.add(PageManager()..setPlugin(plugin));
+
+    return copyWith(newIndex: pages - 1, pageManagers: [..._pageManagers]);
+  }
+
+  TabsState closeView(String pluginId) {
+    _pageManagers.removeWhere((pm) => pm.plugin.id == pluginId);
+
+    final newIndex = currentIndex > pages - 1 && currentIndex > 0
```
Comment (LucasXu0) [inline frontend/appflowy_flutter/lib/workspace/application/tabs/tabs_state.dart:36]: Please add some comments indicating why we need the `currentIndex > 0`.

File: frontend/appflowy_flutter/lib/workspace/application/tabs/tabs_state.dart [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+  PageManager get currentPageManager => _pageManagers[currentIndex];
+  List<PageManager> get pageManagers => _pageManagers;
+
+  TabsState({
+    this.currentIndex = 0,
+    List<PageManager>? pageManagers,
+  }) : _pageManagers = pageManagers ?? [PageManager()];
+
+  /// This opens a new tab, only if the id of the plugin
+  /// being opened, is disimilar to all currently open tabs.
```
[the remark this answers is not in the document]
Comment (Xazin) [reply frontend/appflowy_flutter/lib/workspace/application/tabs/tabs_state.dart:17]: If the plugin is currently not opened in an existing tab

File: frontend/appflowy_flutter/lib/workspace/presentation/home/tabs/flowy_tab.dart [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+        onEnter: (_) => _setHovering(true),
+        onExit: (_) => _setHovering(),
+        child: Container(
+          width: 200,
+          height: HomeSizes.tabBarHeigth,
+          decoration: BoxDecoration(
+            color: _getBackgroundColor(),
+          ),
+          child: Container(
+            width: 200,
```
[the remark this answers is not in the document]
Comment (Xazin) [reply frontend/appflowy_flutter/lib/workspace/presentation/home/tabs/flowy_tab.dart:41]: Mistake when refactoring, will resolve and abstract `200`.

**Comment by Xazin** [comment]
@LucasXu0 I found one issue, and I'm not sure what is the best way to handle it.

When the SelectionMenu overlay is present, you can still change the tab, and it doesn't close. This is less than ideal.

What do you believe to be the best approach to forcefully close it?

**Comment by LucasXu0** [comment]
@Xazin Let me see. I need to check out your branch to take a look.

**Review by LucasXu0 — APPROVED** [review]

**Comment by emmggi** [comment]
Once again thank you for implementing this. In what version can we expect this?

**Comment by Xazin** [comment]
> Once again thank you for implementing this. In what version can we expect this?

Hey Emmgii, this will be in our next release (0.2.7).

I might be doing a few tweaks to the behavior so that the UX will be more smooth, I expect the refinements to be in 0.2.8.

**Comment by emmggi** [comment]
One important thing to do is to restore tabs that were open prior to closing the app.

**Comment by Xazin** [comment]
> One important thing to do is to restore tabs that were open prior to closing the app.

Good catch

**Pull request merged**